File: go/rotation.go
package logger

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// RotationOptions configures log file rotation.
type RotationOptions struct {
	Path           string // Directory for log files (default: ".smooai-logs")
	FilenamePrefix string // Prefix for log filenames (default: "output")
	Extension      string // File extension (default: "ansi")
	Size           string // Max file size before rotation (e.g., "1M", "10K")
	Interval       string // Rotation interval (e.g., "1d", "2h")
	MaxFiles       int    // Max rotated files to keep (default: 30)
	MaxTotalSize   string // Max total size of all log files (e.g., "100M")
}

// DefaultRotationOptions returns the default rotation configuration.
func DefaultRotationOptions() RotationOptions {
	return RotationOptions{
		Path:           ".smooai-logs",
		FilenamePrefix: "output",
		Extension:      "ansi",
		Size:           "1M",
		Interval:       "1d",
		MaxFiles:       30,
		MaxTotalSize:   "100M",
	}
}

type rotatingWriter struct {
	opts           RotationOptions
	mu             sync.Mutex
	file           *os.File
	bytesWritten   int64
	maxBytes       int64
	maxTotalBytes  int64
	interval       time.Duration
	currentDir     string
	currentPath    string
	index          int
	intervalAnchor time.Time
}

func newRotatingWriter(opts RotationOptions) (*rotatingWriter, error) {
	maxBytes := parseSize(opts.Size)
	maxTotalBytes := parseSize(opts.MaxTotalSize)
	interval := parseDuration(opts.Interval)

	now := time.Now().UTC()
	dir := logDirectory(opts, now)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, fmt.Errorf("create log directory: %w", err)
	}

	path := filepath.Join(dir, logFilename(opts, now, 0))
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return nil, fmt.Errorf("open log file: %w", err)
	}

	info, _ := f.Stat()
	var written int64
	if info != nil {
		written = info.Size()
	}

	return &rotatingWriter{
		opts:           opts,
		file:           f,
		bytesWritten:   written,
		maxBytes:       maxBytes,
		maxTotalBytes:  maxTotalBytes,
		interval:       interval,
		currentDir:     dir,
		currentPath:    path,
		index:          0,
		intervalAnchor: now,
	}, nil
}

func (w *rotatingWriter) write(data []byte) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	now := time.Now().UTC()
	if w.shouldRotate(now, int64(len(data))) {
		if err := w.rotate(now); err != nil {
			return err
		}
	}

	n, err := w.file.Write(data)
	w.bytesWritten += int64(n)
	if err != nil {
		return err
	}
	return w.file.Sync()
}

func (w *rotatingWriter) shouldRotate(now time.Time, additional int64) bool {
	if w.maxBytes > 0 && w.bytesWritten+additional > w.maxBytes {
		return true
	}
	if w.interval > 0 && now.Sub(w.intervalAnchor) >= w.interval {
		return true
	}
	return false
}

func (w *rotatingWriter) rotate(now time.Time) error {
	if w.file != nil {
		_ = w.file.Close()
	}

	nextIndex := w.index + 1
	newDir := logDirectory(w.opts, now)
	if newDir != w.currentDir {
		nextIndex = 0
	}

	if err := os.MkdirAll(newDir, 0o755); err != nil {
		return fmt.Errorf("create log directory: %w", err)
	}

	path := filepath.Join(newDir, logFilename(w.opts, now, nextIndex))
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("open rotated log file: %w", err)
	}

	w.file = f
	w.bytesWritten = 0
	w.currentDir = newDir
	w.currentPath = path
	w.index = nextIndex
	w.intervalAnchor = now

	w.enforceLimits()
	return nil
}

func (w *rotatingWriter) enforceLimits() {
	entries, err := os.ReadDir(w.currentDir)
	if err != nil {
		return
	}

	type fileInfo struct {
		path    string
		modTime time.Time
		size    int64
	}

	var files []fileInfo
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		if !strings.HasPrefix(name, w.opts.FilenamePrefix) || !strings.HasSuffix(name, w.opts.Extension) {
			continue
		}
		info, err := entry.Info()
		if err != nil {
			continue
		}
		files = append(files, fileInfo{
			path:    filepath.Join(w.currentDir, name),
			modTime: info.ModTime(),
			size:    info.Size(),
		})
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].modTime.Before(files[j].modTime)
	})

	// Enforce max file count.
	for len(files) > w.opts.MaxFiles && len(files) > 0 {
		_ = os.Remove(files[0].path)
		files = files[1:]
	}

	// Enforce max total size.
	if w.maxTotalBytes > 0 {
		var total int64
		for _, f := range files {
			total += f.size
		}
		for total > w.maxTotalBytes && len(files) > 0 {
			total -= files[0].size
			_ = os.Remove(files[0].path)
			files = files[1:]
		}
	}
}

func (w *rotatingWriter) close() error {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.file != nil {
		return w.file.Close()
	}
	return nil
}

func logDirectory(opts RotationOptions, t time.Time) string {
	folder := fmt.Sprintf("%04d-%02d", t.Year(), t.Month())
	return filepath.Join(opts.Path, folder)
}

func logFilename(opts RotationOptions, t time.Time, index int) string {
	return fmt.Sprintf("%s-%04d-%02d-%02d-%03d.%s",
		opts.FilenamePrefix,
		t.Year(), t.Month(), t.Day(),
		index,
		opts.Extension,
	)
}

func parseSize(s string) int64 {
	s = strings.TrimSpace(strings.ToUpper(s))
	if s == "" {
		return 0
	}
	if strings.HasSuffix(s, "K") {
		n, _ := strconv.ParseInt(s[:len(s)-1], 10, 64)
		return n * 1024
	}
	if strings.HasSuffix(s, "M") {
		n, _ := strconv.ParseInt(s[:len(s)-1], 10, 64)
		return n * 1024 * 1024
	}
	if strings.HasSuffix(s, "G") {
		n, _ := strconv.ParseInt(s[:len(s)-1], 10, 64)
		return n * 1024 * 1024 * 1024
	}
	n, _ := strconv.ParseInt(s, 10, 64)
	return n
}

func parseDuration(s string) time.Duration {
	s = strings.TrimSpace(strings.ToLower(s))
	if s == "" {
		return 0
	}
	if strings.HasSuffix(s, "s") {
		n, _ := strconv.Atoi(s[:len(s)-1])
		return time.Duration(n) * time.Second
	}
	if strings.HasSuffix(s, "m") {
		n, _ := strconv.Atoi(s[:len(s)-1])
		return time.Duration(n) * time.Minute
	}
	if strings.HasSuffix(s, "h") {
		n, _ := strconv.Atoi(s[:len(s)-1])
		return time.Duration(n) * time.Hour
	}
	if strings.HasSuffix(s, "d") {
		n, _ := strconv.Atoi(s[:len(s)-1])
		return time.Duration(n) * 24 * time.Hour
	}
	if strings.HasSuffix(s, "w") {
		n, _ := strconv.Atoi(s[:len(s)-1])
		return time.Duration(n) * 7 * 24 * time.Hour
	}
	return 0
}

File: rust/logger/src/rotation.rs
use std::fs::{self, File, OpenOptions};
use std::io::{self, Write};
use std::path::{Path, PathBuf};
use std::time::UNIX_EPOCH;

use chrono::{Datelike, Duration, Utc};
use parking_lot::Mutex;

#[derive(Clone, Debug)]
pub struct RotationOptions {
    pub path: PathBuf,
    pub filename_prefix: String,
    pub extension: String,
    pub size: Option<String>,
    pub interval: Option<String>,
    pub max_files: usize,
    pub max_total_size: Option<String>,
}

impl Default for RotationOptions {
    fn default() -> Self {
        Self {
            path: PathBuf::from(".smooai-logs"),
            filename_prefix: "output".into(),
            extension: "ansi".into(),
            size: Some("1M".into()),
            interval: Some("1d".into()),
            max_files: 30,
            max_total_size: Some("100M".into()),
        }
    }
}

#[derive(Debug)]
struct WriterState {
    file: File,
    bytes_written: u64,
    current_dir: PathBuf,
    current_path: PathBuf,
    index: u32,
    interval_anchor: chrono::DateTime<Utc>,
}

#[derive(Debug)]
pub struct RotatingFileWriter {
    options: RotationOptions,
    max_bytes: Option<u64>,
    max_total_bytes: Option<u64>,
    interval: Option<Duration>,
    state: Mutex<WriterState>,
}

impl RotatingFileWriter {
    pub fn new(options: RotationOptions) -> io::Result<Self> {
        let max_bytes = options.size.as_ref().and_then(|s| parse_size(s).ok());
        let max_total_bytes = options.max_total_size.as_ref().and_then(|s| parse_size(s).ok());
        let interval = options.interval.as_ref().and_then(|s| parse_interval(s).ok());

        let now = Utc::now();
        let (file, current_dir, current_path) = open_file(&options, &now, 0)?;
        let bytes_written = file.metadata().map(|m| m.len()).unwrap_or(0);

        Ok(Self {
            options,
            max_bytes,
            max_total_bytes,
            interval,
            state: Mutex::new(WriterState {
                file,
                bytes_written,
                current_dir,
                current_path,
                index: 0,
                interval_anchor: now,
            }),
        })
    }

    pub fn write(&self, payload: &str) -> io::Result<()> {
        let mut state = self.state.lock();
        let now = Utc::now();
        let payload_bytes = payload.as_bytes();
        if self.should_rotate(&state, &now, payload_bytes.len() as u64) {
            rotate(&self.options, &mut state, &now, self.max_total_bytes)?;
        }

        state.file.write_all(payload_bytes)?;
        state.bytes_written += payload_bytes.len() as u64;
        state.file.flush()
    }

    fn should_rotate(&self, state: &WriterState, now: &chrono::DateTime<Utc>, additional: u64) -> bool {
        if let Some(max_bytes) = self.max_bytes {
            if state.bytes_written + additional > max_bytes {
                return true;
            }
        }

        if let Some(interval) = self.interval {
            if *now - state.interval_anchor >= interval {
                return true;
            }
        }

        false
    }
}

fn rotate(options: &RotationOptions, state: &mut WriterState, now: &chrono::DateTime<Utc>, max_total_bytes: Option<u64>) -> io::Result<()> {
    let mut next_index = state.index + 1;
    let current_dir = log_directory(options, now);
    if current_dir != state.current_dir {
        next_index = 0;
    }

    let (file, dir, path) = open_file(options, now, next_index)?;

    state.file = file;
    state.bytes_written = 0;
    state.current_dir = dir.clone();
    state.current_path = path.clone();
    state.index = next_index;
    state.interval_anchor = *now;

    enforce_limits(options, &dir, max_total_bytes)
}

fn enforce_limits(options: &RotationOptions, directory: &Path, max_total_bytes: Option<u64>) -> io::Result<()> {
    if !directory.exists() {
        return Ok(());
    }

    let mut entries: Vec<_> = fs::read_dir(directory)?
        .filter_map(|entry| entry.ok())
        .filter(|entry| entry.file_type().map(|ft| ft.is_file()).unwrap_or(false))
        .filter(|entry| has_prefix(entry.file_name(), &options.filename_prefix, &options.extension))
        .collect();

    entries.sort_by(|a, b| {
        let a_time = a
            .metadata()
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
            .unwrap_or_default();
        let b_time = b
            .metadata()
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
            .unwrap_or_default();
        a_time.cmp(&b_time)
    });

    while entries.len() > options.max_files {
        if let Some(entry) = entries.first() {
            let _ = fs::remove_file(entry.path());
        }
        entries.remove(0);
    }

    if let Some(limit) = max_total_bytes {
        let mut total: u64 = entries.iter().filter_map(|entry| entry.metadata().ok().map(|m| m.len())).sum();
        while total > limit && !entries.is_empty() {
            if let Some(entry) = entries.first() {
                let path = entry.path();
                let size = entry.metadata().ok().map(|m| m.len()).unwrap_or(0);
                let _ = fs::remove_file(&path);
                total = total.saturating_sub(size);
            }
            entries.remove(0);
        }
    }

    Ok(())
}

fn has_prefix(name: std::ffi::OsString, prefix: &str, extension: &str) -> bool {
    let name = name.to_string_lossy();
    name.starts_with(prefix) && name.ends_with(extension)
}

fn open_file(options: &RotationOptions, now: &chrono::DateTime<Utc>, index: u32) -> io::Result<(File, PathBuf, PathBuf)> {
    let directory = log_directory(options, now);
    fs::create_dir_all(&directory)?;
    let filename = log_filename(options, now, index);
    let path = directory.join(filename);
    let file = OpenOptions::new().create(true).append(true).open(&path)?;
    Ok((file, directory, path))
}

fn log_directory(options: &RotationOptions, now: &chrono::DateTime<Utc>) -> PathBuf {
    let folder = format!("{:04}-{:02}", now.year(), now.month());
    options.path.join(folder)
}

fn log_filename(options: &RotationOptions, now: &chrono::DateTime<Utc>, index: u32) -> String {
    format!(
        "{}-{:04}-{:02}-{:02}-{:03}.{}",
        options.filename_prefix,
        now.year(),
        now.month(),
        now.day(),
        index,
        options.extension
    )
}

fn parse_size(size: &str) -> Result<u64, &'static str> {
    let upper = size.trim().to_uppercase();
    if let Some(stripped) = upper.strip_suffix('K') {
        return stripped.parse::<u64>().map(|n| n * 1024).map_err(|_| "invalid size");
    }
    if let Some(stripped) = upper.strip_suffix('M') {
        return stripped.parse::<u64>().map(|n| n * 1024 * 1024).map_err(|_| "invalid size");
    }
    if let Some(stripped) = upper.strip_suffix('G') {
        return stripped.parse::<u64>().map(|n| n * 1024 * 1024 * 1024).map_err(|_| "invalid size");
    }
    upper.parse::<u64>().map_err(|_| "invalid size")
}

fn parse_interval(interval: &str) -> Result<Duration, &'static str> {
    let lower = interval.trim().to_lowercase();
    if let Some(stripped) = lower.strip_suffix('s') {
        return stripped.parse::<i64>().map(Duration::seconds).map_err(|_| "invalid interval");
    }
    if let Some(stripped) = lower.strip_suffix('m') {
        return stripped.parse::<i64>().map(Duration::minutes).map_err(|_| "invalid interval");
    }
    if let Some(stripped) = lower.strip_suffix('h') {
        return stripped.parse::<i64>().map(Duration::hours).map_err(|_| "invalid interval");
    }
    if let Some(stripped) = lower.strip_suffix('d') {
        return stripped.parse::<i64>().map(Duration::days).map_err(|_| "invalid interval");
    }
    if let Some(stripped) = lower.strip_suffix('w') {
        return stripped.parse::<i64>().map(|weeks| Duration::days(7 * weeks)).map_err(|_| "invalid interval");
    }
    Err("invalid interval")
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn rotation_options_defaults() {
        let options = RotationOptions::default();
        assert_eq!(options.filename_prefix, "output");
    }

    #[test]
    fn rotating_writer_creates_file() {
        let dir = tempdir().unwrap();
        let options = RotationOptions {
            path: dir.path().into(),
            ..Default::default()
        };
        let writer = RotatingFileWriter::new(options).unwrap();
        writer.write("test line\n").unwrap();
    }
}
